File: backend/ride-share/src/main/java/mk/ukim/finki/rideshare/service/RideService.java
package mk.ukim.finki.rideshare.service;

import mk.ukim.finki.rideshare.model.Ride;
import mk.ukim.finki.rideshare.model.User;
import org.springframework.data.domain.Sort;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.UUID;

public interface RideService {

    List<Ride> search(String origin, String destination, LocalDate date, Integer seats, Sort sort);

    List<Ride> getAllForActiveUser(Boolean includePast);

    Boolean hasRideEnoughSeatsLeft(Ride ride, Integer seats);

    Ride getById(Long id);

    Ride getByUuid(UUID uuid);

    Ride create(String origin,
                Double originLatitude,
                Double originLongitude,
                String destination,
                Double destinationLatitude,
                Double destinationLongitude,
                Boolean isDoorToDoor,
                ZonedDateTime departureTime,
                Boolean isDepartureTimeFlexible,
                Double price,
                Boolean hasLuggageSpace,
                Integer capacity,
                Boolean isInstantBookingEnabled);

    List<Ride> getAllByOriginLikeAndDestinationLike(String origin, String destination);

    Ride cancel(Long rideId);

    Long getNumberOfCancellationsByProviderInTheLastMonth(User provider);
}

File: frontend/ride-share/src/services/rideService.ts
import axiosConfig from "../config/axiosConfig";
import { CreateRideRequest } from "../interfaces/request/CreateRideRequest";
import { RidePriceStatisticsResponse } from "../interfaces/response/RidePriceStatisticsResponse";
import { RideResponse } from "../interfaces/response/RideResponse";

const path = 'rides';

export const RideService = {
    search: async (origin?: string, destination?: string, date?: string, seats?: number, sortBy?: string, sortDirection?: string) => {
        return axiosConfig.get<RideResponse[]>(`${path}/search`, {
            params: {
                origin,
                destination,
                date,
                seats,
                sortBy,
                sortDirection
            }
        });
    },
    getAllForActiveUser: async (includePast: boolean) => {
        return axiosConfig.get<RideResponse[]>(path, {
            params: { includePast }
        });
    },
    getById: async (id: number) => {
        return axiosConfig.get<RideResponse>(`${path}/${id}`);
    },
    getByUuid: async(uuid: string) => {
        return axiosConfig.get<RideResponse>(`${path}/uuid/${uuid}`);
    },
    create: async (request: CreateRideRequest) => {
        return axiosConfig.post(path, request);
    },
    getPriceStatisticsForRide: async (origin: string, destination: string) => {
        return axiosConfig.get<RidePriceStatisticsResponse>(`${path}/recommended-price`, { params: { origin, destination }});
    }
}
